README.md
# HTML5 QR-code reader - no flash, only html and javascript

Actual version: 1.1.0


A very basic QR-code reader based on:
- jQuery: http://jquery.com/
- webcamjs: https://github.com/jhuckaby/webcamjs
- QR-code library: https://github.com/LazarSoft/jsqrcode


The example (index.html) uses webcamjs to read an image 
from the webcam, please visit [the docs of webcamjs](https://github.com/jhuckaby/webcamjs/blob/master/DOCS.md) if there are any questions.


Examples: 
- [Demo](https://html5qrcode.ephra.im/) -> example with webcamjs


Usage:
Only clone or download this script and test it on your local webserver.
Make sure you're using **HTTPS-protocol** to get it work with modern browsers.
Enjoy!


Other projects with webcam QR code reading:
- [instascan](https://github.com/schmich/instascan)


changelog.txt
18.08.2019:
- updated jQuery to v3.4.1
- removed photobooth and say-cheese
- added webcamjs to get access to the webcam
- release of version 1.1.0


02.03.2017:
- updated jQuery to v3.1.1
- added autoplay option to say cheese
- minified included scripts


18.05.2014:
- replaced linkify function in effects_saycheese.js with a better detection


10.05.2014:
- added reload button
- added link detection for index_saycheese.html


js/qr/qrcode.js
/*
   Copyright 2011 Lazar Laszlo (lazarsoft@gmail.com, www.lazarsoft.info)
   
   Licensed under the Apache License, Version 2.0 (the "License");
   you may not use this file except in compliance with the License.
   You may obtain a copy of the License at

       http://www.apache.org/licenses/LICENSE-2.0

   Unless required by applicable law or agreed to in writing, software
   distributed under the License is distributed on an "AS IS" BASIS,
   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
   See the License for the specific language governing permissions and
   limitations under the License.
*/


qrcode = {};
qrcode.imagedata = null;
qrcode.width = 0;
qrcode.height = 0;
qrcode.qrCodeSymbol = null;
qrcode.debug = false;
qrcode.maxImgSize = 1024*1024;

qrcode.sizeOfDataLengthInfo =  [  [ 10, 9, 8, 8 ],  [ 12, 11, 16, 10 ],  [ 14, 13, 16, 12 ] ];

qrcode.callback = null;

qrcode.decode = function(src){
    
    if(arguments.length==0)
    {
        var canvas_qr = document.getElementById("qr-canvas");
        var context = canvas_qr.getContext('2d');
        qrcode.width = canvas_qr.width;
        qrcode.height = canvas_qr.height;
        qrcode.imagedata = context.getImageData(0, 0, qrcode.width, qrcode.height);
        qrcode.result = qrcode.process(context);
        if(qrcode.callback!=null)
            qrcode.callback(qrcode.result);
        return qrcode.result;
    }
    else
    {
        var image = new Image();
        image.onload=function(){
            //var canvas_qr = document.getElementById("qr-canvas");
            var canvas_qr = document.createElement('canvas');
            var context = canvas_qr.getContext('2d');
            var nheight = image.height;
            var nwidth = image.width;
            if(image.width*image.height>qrcode.maxImgSize)
            {
                var ir = image.width / image.height;
                nheight = Math.sqrt(qrcode.maxImgSize/ir);
                nwidth=ir*nheight;
            }

            canvas_qr.width = nwidth;
            canvas_qr.height = nheight;
            
            context.drawImage(image, 0, 0, canvas_qr.width, canvas_qr.height );
            qrcode.width = canvas_qr.width;
            qrcode.height = canvas_qr.height;
            try{
                qrcode.imagedata = context.getImageData(0, 0, canvas_qr.width, canvas_qr.height);
            }catch(e){
                qrcode.result = "Cross domain image reading not supported in your browser! Save it to your computer then drag and drop the file!";
                if(qrcode.callback!=null)
                    qrcode.callback(qrcode.result);
                return;
            }
            
            try
            {
                qrcode.result = qrcode.process(context);
            }
            catch(e)
            {
                //console.log(e);
                qrcode.result = "error decoding QR Code";
            }
            if(qrcode.callback!==null)
                qrcode.callback(qrcode.result);
        }
        image.src = src;
    }
}

qrcode.isUrl = function(s)
{
    var regexp = /(ftp|http|https):\/\/(\w+:{0,1}\w*@)?(\S+)(:[0-9]+)?(\/|\/([\w#!:.?+=&%@!\-\/]))?/;
    return regexp.test(s);
}

qrcode.decode_url = function (s)
{
  var escaped = "";
  try{
    escaped = escape( s );
  }
  catch(e)
  {
    console.log(e);
    escaped = s;
  }
  var ret = "";
  try{
    ret = decodeURIComponent( escaped );
  }
  catch(e)
  {
    console.log(e);
    ret = escaped;
  }
  return ret;
}

qrcode.decode_utf8 = function ( s )
{
    if(qrcode.isUrl(s))
        return qrcode.decode_url(s);
    else
        return s;
}

qrcode.process = function(ctx){
    
    var start = new Date().getTime();

    var image = qrcode.grayScaleToBitmap(qrcode.grayscale());
    //var image = qrcode.binarize(128);
    
    if(qrcode.debug)
    {
        for (var y = 0; y < qrcode.height; y++)
        {
            for (var x = 0; x < qrcode.width; x++)
            {
                var point = (x * 4) + (y * qrcode.width * 4);
                qrcode.imagedata.data[point] = image[x+y*qrcode.width]?0:0;
                qrcode.imagedata.data[point+1] = image[x+y*qrcode.width]?0:0;
                qrcode.imagedata.data[point+2] = image[x+y*qrcode.width]?255:0;
            }
        }
        ctx.putImageData(qrcode.imagedata, 0, 0);
    }
    
    //var finderPatternInfo = new FinderPatternFinder().findFinderPattern(image);
    
    var detector = new Detector(image);

    var qRCodeMatrix = detector.detect();
    
    /*for (var y = 0; y < qRCodeMatrix.bits.Height; y++)
    {
        for (var x = 0; x < qRCodeMatrix.bits.Width; x++)
        {
            var point = (x * 4*2) + (y*2 * qrcode.width * 4);
            qrcode.imagedata.data[point] = qRCodeMatrix.bits.get_Renamed(x,y)?0:0;
            qrcode.imagedata.data[point+1] = qRCodeMatrix.bits.get_Renamed(x,y)?0:0;
            qrcode.imagedata.data[point+2] = qRCodeMatrix.bits.get_Renamed(x,y)?255:0;
        }
    }*/
    if(qrcode.debug)
        ctx.putImageData(qrcode.imagedata, 0, 0);
    
    var reader = Decoder.decode(qRCodeMatrix.bits);
    var data = reader.DataByte;
    var str="";
    for(var i=0;i<data.length;i++)
    {
        for(var j=0;j<data[i].length;j++)
            str+=String.fromCharCode(data[i][j]);
    }
    
    var end = new Date().getTime();
    var time = end - start;
    //console.log(time);
    
    return qrcode.decode_utf8(str);
    //alert("Time:" + time + " Code: "+str);
}

qrcode.getPixel = function(x,y){
    if (qrcode.width < x) {
        throw "point error";
    }
    if (qrcode.height < y) {
        throw "point error";
    }
    point = (x * 4) + (y * qrcode.width * 4);
    p = (qrcode.imagedata.data[point]*33 + qrcode.imagedata.data[point + 1]*34 + qrcode.imagedata.data[point + 2]*33)/100;
    return p;
}

qrcode.binarize = function(th){
    var ret = new Array(qrcode.width*qrcode.height);
    for (var y = 0; y < qrcode.height; y++)
    {
        for (var x = 0; x < qrcode.width; x++)
        {
            var gray = qrcode.getPixel(x, y);
            
            ret[x+y*qrcode.width] = gray<=th?true:false;
        }
    }
    return ret;
}

qrcode.getMiddleBrightnessPerArea=function(image)
{
    var numSqrtArea = 4;
    //obtain middle brightness((min + max) / 2) per area
    var areaWidth = Math.floor(qrcode.width / numSqrtArea);
    var areaHeight = Math.floor(qrcode.height / numSqrtArea);
    var minmax = new Array(numSqrtArea);
    for (var i = 0; i < numSqrtArea; i++)
    {
        minmax[i] = new Array(numSqrtArea);
        for (var i2 = 0; i2 < numSqrtArea; i2++)
        {
            minmax[i][i2] = new Array(0,0);
        }
    }
    for (var ay = 0; ay < numSqrtArea; ay++)
    {
        for (var ax = 0; ax < numSqrtArea; ax++)
        {
            minmax[ax][ay][0] = 0xFF;
            for (var dy = 0; dy < areaHeight; dy++)
            {
                for (var dx = 0; dx < areaWidth; dx++)
                {
                    var target = image[areaWidth * ax + dx+(areaHeight * ay + dy)*qrcode.width];
                    if (target < minmax[ax][ay][0])
                        minmax[ax][ay][0] = target;
                    if (target > minmax[ax][ay][1])
                        minmax[ax][ay][1] = target;
                }
            }
            //minmax[ax][ay][0] = (minmax[ax][ay][0] + minmax[ax][ay][1]) / 2;
        }
    }
    var middle = new Array(numSqrtArea);
    for (var i3 = 0; i3 < numSqrtArea; i3++)
    {
        middle[i3] = new Array(numSqrtArea);
    }
    for (var ay = 0; ay < numSqrtArea; ay++)
    {
        for (var ax = 0; ax < numSqrtArea; ax++)
        {
            middle[ax][ay] = Math.floor((minmax[ax][ay][0] + minmax[ax][ay][1]) / 2);
            //Console.out.print(middle[ax][ay] + ",");
        }
        //Console.out.println("");
    }
    //Console.out.println("");
    
    return middle;
}

qrcode.grayScaleToBitmap=function(grayScale)
{
    var middle = qrcode.getMiddleBrightnessPerArea(grayScale);
    var sqrtNumArea = middle.length;
    var areaWidth = Math.floor(qrcode.width / sqrtNumArea);
    var areaHeight = Math.floor(qrcode.height / sqrtNumArea);
    var bitmap = new Array(qrcode.height*qrcode.width);
    
    for (var ay = 0; ay < sqrtNumArea; ay++)
    {
        for (var ax = 0; ax < sqrtNumArea; ax++)
        {
            for (var dy = 0; dy < areaHeight; dy++)
            {
                for (var dx = 0; dx < areaWidth; dx++)
                {
                    bitmap[areaWidth * ax + dx+ (areaHeight * ay + dy)*qrcode.width] = (grayScale[areaWidth * ax + dx+ (areaHeight * ay + dy)*qrcode.width] < middle[ax][ay])?true:false;
                }
            }
        }
    }
    return bitmap;
}

qrcode.grayscale = function(){
    var ret = new Array(qrcode.width*qrcode.height);
    for (var y = 0; y < qrcode.height; y++)
    {
        for (var x = 0; x < qrcode.width; x++)
        {
            var gray = qrcode.getPixel(x, y);
            
            ret[x+y*qrcode.width] = gray;
        }
    }
    return ret;
}




function URShift( number,  bits)
{
    if (number >= 0)
        return number >> bits;
    else
        return (number >> bits) + (2 << ~bits);
}


Array.prototype.remove = function(from, to) {
  var rest = this.slice((to || from) + 1 || this.length);
  this.length = from < 0 ? this.length + from : from;
  return this.push.apply(this, rest);
};


js/saycheese.js
/*
 * Say Cheese!
 * Lee Machin, 2012
 * http://leemach.in, http://new-bamboo.co.uk
 *
 * Minimal javascript library for integrating a webcam and snapshots into your app.
 *
 * Handles starting up the webcam and rendering the element, and also capturing shots
 * in a separate canvas element.
 *
 * Depends on video and canvas, and of course, getUserMedia. It's unlikely to work
 * on anything but the newest browsers.
 */

var SayCheese = (function() {

  var SayCheese;

  navigator.getUserMedia = (navigator.getUserMedia ||
                            navigator.webkitGetUserMedia ||
                            navigator.mozGetUserMedia ||
                            navigator.msGetUserMedia ||
                            false);

  window.AudioContext = (window.AudioContext ||
                         window.webkitAudioContext);

  window.URL = (window.URL ||
                window.webkitURL);

  SayCheese = function SayCheese(element, options) {
    this.snapshots = [],
    this.video = null,
    this.events = {},
    this.stream = null,
    this.options = {
      snapshots: true,
      audio: false
    };

    this.setOptions(options);
    this.element = document.querySelector(element);
    return this;
  };

  SayCheese.prototype.on = function on(evt, handler) {
    if (this.events.hasOwnProperty(evt) === false) {
      this.events[evt] = [];
    }

    this.events[evt].push(handler)
  };

  SayCheese.prototype.off = function off(evt, handler) {
    this.events[evt] = this.events[evt].filter(function(h) {
      return h !== handler;
    });
  };

  SayCheese.prototype.trigger = function trigger(evt, data) {
    if (this.events.hasOwnProperty(evt) === false) {
      return false;
    }

    this.events[evt].forEach(function(handler) {
      handler.call(this, data);
    }.bind(this));
  };

  SayCheese.prototype.setOptions = function setOptions(options) {
    // just use naïve, shallow cloning
    for (var opt in options) {
      this.options[opt] = options[opt];
    }
  }

  SayCheese.prototype.getStreamUrl = function getStreamUrl() {
    if (window.URL && window.URL.createObjectURL) {
      return window.URL.createObjectURL(this.stream);
    } else {
      return this.stream;
    }
  };

  SayCheese.prototype.createVideo = function createVideo() {
    var width     = 320,
        height    = 0,
        streaming = false;

    this.video = document.createElement('video');

    this.video.addEventListener('canplay', function() {
      if (!streaming) {
        height = this.video.videoHeight / (this.video.videoWidth / width);
        this.video.style.width = width;
        this.video.style.height = height;
        streaming = true;
        return this.trigger('start');
      }
    }.bind(this), false);
  };

  SayCheese.prototype.linkAudio = function linkAudio() {
    this.audioCtx = new window.AudioContext();
    this.audioStream = this.audioCtx.createMediaStreamSource(this.stream);

    var biquadFilter = this.audioCtx.createBiquadFilter();

    this.audioStream.connect(biquadFilter);
    biquadFilter.connect(this.audioCtx.destination);
  };

  SayCheese.prototype.takeSnapshot = function takeSnapshot(width, height) {
    if (this.options.snapshots === false) {
      return false;
    }

    width  = width || this.video.videoWidth;
    height = height || this.video.videoHeight;

    var snapshot = document.createElement('canvas'),
        ctx      = snapshot.getContext('2d');

    snapshot.width  = width;
    snapshot.height = height;

    ctx.drawImage(this.video, 0, 0, width, height);

    this.snapshots.push(snapshot);
    this.trigger('snapshot', snapshot);

    ctx = null;
  };

  /* Start up the stream, if possible */
  SayCheese.prototype.start = function start() {

    // fail fast and softly if browser not supported
    if (navigator.getUserMedia === false) {
      this.trigger('error', 'NOT_SUPPORTED');
      return false;
    }

    var success = function success(stream) {
      this.stream = stream;
      this.createVideo();

      if (navigator.mozGetUserMedia) {
        this.video.mozSrcObject = stream;
      } else {
        this.video.src = this.getStreamUrl();
      }

      if (this.options.audio === true) {
        try {
          this.linkAudio();
        } catch(e) {
          this.trigger('error', 'AUDIO_NOT_SUPPORTED');
        }
      }

      this.element.appendChild(this.video);
      this.video.play();
    }.bind(this);

    /* error is also called when someone denies access */
    var error = function error(error) {
      this.trigger('error', error);
    }.bind(this);

    return navigator.getUserMedia({ video: true, audio: this.options.audio }, success, error);
  };

  SayCheese.prototype.stop = function stop() {
    this.stream.stop();

    if (window.URL && window.URL.revokeObjectURL) {
      window.URL.revokeObjectURL(this.video.src);
    }

    return this.trigger('stop');
  };

  return SayCheese;

})();



js/say-cheese.js
/*
 * Say Cheese!
 * Lee Machin, 2012
 * http://leemach.in, http://new-bamboo.co.uk
 *
 * Minimal javascript library for integrating a webcam and snapshots into your app.
 *
 * Handles starting up the webcam and rendering the element, and also capturing shots
 * in a separate canvas element.
 *
 * Depends on video and canvas, and of course, getUserMedia. It's unlikely to work
 * on anything but the newest browsers.
 * 
 * Small updates: 02.03.2017
 */

var SayCheese = (function () {

	var SayCheese;

	navigator.getUserMedia = (navigator.getUserMedia ||
			navigator.webkitGetUserMedia ||
			navigator.mozGetUserMedia ||
			navigator.msGetUserMedia ||
			false);

	window.AudioContext = (window.AudioContext ||
			window.webkitAudioContext);

	window.URL = (window.URL ||
			window.webkitURL);

	var ERRORS = {
		NOT_SUPPORTED: 'NOT_SUPPORTED',
		AUDIO_NOT_SUPPORTED: 'AUDIO_NOT_SUPPORTED'
	};

	SayCheese = function SayCheese(element, options) {
		this.snapshots = [],
				this.video = null,
				this.events = {},
				this.stream = null,
				this.options = {
					videoSource: null,
					snapshots: true,
					audio: false,
					width: 320
				};

		this.setOptions(options);
		this.element = document.querySelector(element);
		//console.log('Initialized');
		return this;
	};

	SayCheese.prototype.on = function on(evt, handler) {
		if (this.events.hasOwnProperty(evt) === false) {
			this.events[evt] = [];
		}

		this.events[evt].push(handler);
	};

	SayCheese.prototype.off = function off(evt, handler) {
		this.events[evt] = this.events[evt].filter(function (h) {
			return h !== handler;
		});
	};

	SayCheese.prototype.trigger = function trigger(evt, data) {
		//console.log('triggered!');
		if (this.events.hasOwnProperty(evt) === false) {
			return false;
		}

		this.events[evt].forEach(function (handler) {
			handler.call(this, data);
		}.bind(this));
	};

	SayCheese.prototype.setOptions = function setOptions(options) {
		// just use naïve, shallow cloning
		for (var opt in options) {
			this.options[opt] = options[opt];
		}
	};

	SayCheese.prototype.getStreamUrl = function getStreamUrl() {
		if (window.URL && window.URL.createObjectURL) {
			return window.URL.createObjectURL(this.stream);
		} else {
			return this.stream;
		}
	};

	SayCheese.prototype.createVideo = function createVideo() {
		//console.log('creating video');
		var width = this.options.width,
				height = 0,
				streaming = false;

		this.video = document.createElement('video');

		this.video.addEventListener('canplay', function () {
			if (!streaming) {
				height = this.video.videoHeight / (this.video.videoWidth / width);
				this.video.width = width;
				this.video.height = height;
				this.video.setAttribute('autoplay','');
				streaming = true;
				return this.trigger('start');
			}
		}.bind(this), false);
	};

	SayCheese.prototype.linkAudio = function linkAudio() {
		this.audioCtx = new window.AudioContext();
		this.audioStream = this.audioCtx.createMediaStreamSource(this.stream);

		var biquadFilter = this.audioCtx.createBiquadFilter();

		this.audioStream.connect(biquadFilter);
		biquadFilter.connect(this.audioCtx.destination);
	};

	SayCheese.prototype.takeSnapshot = function takeSnapshot(width, height) {
		if (this.options.snapshots === false) {
			return false;
		}

		width = width || this.video.videoWidth;
		height = height || this.video.videoHeight;

		var snapshot = document.createElement('canvas'),
				ctx = snapshot.getContext('2d');

		snapshot.width = width;
		snapshot.height = height;

		ctx.drawImage(this.video, 0, 0, width, height);

		this.snapshots.push(snapshot);
		this.trigger('snapshot', snapshot);

		ctx = null;
	};

	/* Start up the stream, if possible */
	SayCheese.prototype.start = function start() {

		//console.log('starting say-cheese...');

		// fail fast and softly if browser not supported
		if (navigator.getUserMedia === false) {
			//console.log('failing fast and softly...');
			this.trigger('error', ERRORS.NOT_SUPPORTED);
			return false;
		}

		var success = function success(stream) {
			//console.log('sucess!');
			this.stream = stream;
			this.createVideo();

			if (navigator.mozGetUserMedia) {
				this.video.mozSrcObject = stream;
			} else {
				this.video.src = this.getStreamUrl();
			}

			if (this.options.audio === true) {
				try {
					this.linkAudio();
				} catch (e) {
					this.trigger('error', ERRORS.AUDIO_NOT_SUPPORTED);
				}
			}

			this.element.appendChild(this.video);
			this.video.play();

			this.trigger('success');
		}.bind(this);

		/* error is also called when someone denies access */
		var error = function error(error) {
			//console.log('access denied');
			this.trigger('error', error);
		}.bind(this);

		//console.log('Getting user media...');

		return navigator.getUserMedia({video: {
				optional: [{
						sourceId: this.options.videoSource
					}]
			}, audio: this.options.audio}, success, error);
	};

	SayCheese.prototype.stop = function stop() {
		this.stream.stop();

		if (window.URL && window.URL.revokeObjectURL) {
			window.URL.revokeObjectURL(this.video.src);
		}

		return this.trigger('stop');
	};

	return SayCheese;

})();



js/effects_saycheese.js
// user media
// inspired by https://www.simpl.info/getusermedia/sources/index.html

var videoSelect = document.querySelector('select#videoSource');

var gotSources = function (sourceInfos) {
	//console.log('Detecting user media video sources...');
	for (var i = 0; i !== sourceInfos.length; ++i) {
		var sourceInfo = sourceInfos[i];
		var option = document.createElement('option');
		option.value = sourceInfo.id;
		if (sourceInfo.kind === 'video') {
			option.text = sourceInfo.label || 'Camera '
					+ (videoSelect.length);
			videoSelect.appendChild(option);
			//console.log('Camera Source detected: ' + option.text);
		} else {
			//console.log('Other Source detected: ' + (sourceInfo.label || 'other ' + (videoSelect.length + 1)));
		}
	}
};

// detect user media
if (typeof MediaStreamTrack === 'undefined'
		|| typeof MediaStreamTrack.getSources === 'undefined') {
	//console.log('This browser does not support MediaStreamTrack.');
} else {
	//console.log('MediaStreamTrack supported.');
	MediaStreamTrack.getSources(gotSources);
}

var scanner;

function go() {
	//console.log('go!');

	if (scanner) {
		scanner.stop();
	}

	scanner = new SayCheese('#example', {
		snapshots: true,
		videoSource: videoSelect.value
	});

	scanner.on('error', function (error) {
		$('#example').html('<p>This plugin cannot run. Check if your browser blocks it.</p>');
	});

	scanner.on('snapshot', function (snapshot) {
		qrCodeDecoder(snapshot.toDataURL());
	});

	qrcode.callback = showInfo;

	scanner.on('success', function () {
		scanCode(scanner);
	});

	//console.log('starting scanner...');
	scanner.start();
}

// recursive function for scanning code
function scanCode(scanner) {
	//console.log('taking snapshot...');
	scanner.takeSnapshot();
	setTimeout(function () {
		scanCode(scanner);
	}, 1000);
}

// decode the img
function qrCodeDecoder(dataUrl) {
	qrcode.decode(dataUrl);
}

// show info from qr code
function showInfo(data) {
	if (data !== 'error decoding QR Code') {
		var htmldata = linkify(data);
		$("#qrContent").addClass('hasContent');
		$("#qrContent p").html(htmldata);
	} else {
		if ($("#qrContent").hasClass('hasContent')) {
		} else {
			$("#qrContent p").html('No QR Code in sight.');
		}
	}
}

// builds a link if there is an uri or a mail address
function linkify(inputText) {
	var replacedText, replacePattern1, replacePattern2, replacePattern3;

	//URLs starting with http://, https://, or ftp://
	replacePattern1 = /(\b(https?|ftp):\/\/[-A-Z0-9+&@#\/%?=~_|!:,.;]*[-A-Z0-9+&@#\/%=~_|])/gim;
	replacedText = inputText.replace(replacePattern1,
			'<a href="$1" target="_blank">$1</a>');

	//URLs starting with "www." (without // before it, or it'd re-link the ones done above).
	replacePattern2 = /(^|[^\/])(www\.[\S]+(\b|$))/gim;
	replacedText = replacedText.replace(replacePattern2,
			'$1<a href="http://$2" target="_blank">$2</a>');

	//Change email addresses to mailto:: links.
	replacePattern3 = /(([a-zA-Z0-9\-\_\.])+@[a-zA-Z\_]+?(\.[a-zA-Z]{2,6})+)/gim;
	replacedText = replacedText.replace(replacePattern3,
			'<a href="mailto:$1">$1</a>');

	return replacedText;
}


js/photobooth.js
Photobooth = function (container)
{

	var self = this;
	/**
	 * Make it jQuery friendlier
	 */
	if (container.length)
	{
		container = container[ 0 ];
	}

	var fGetUserMedia =
			(
					navigator.getUserMedia ||
					navigator.webkitGetUserMedia ||
					navigator.mozGetUserMedia ||
					navigator.oGetUserMedia ||
					navigator.msieGetUserMedia ||
					false
					);
	/**
	 * Image callback. Will be called when
	 * the user clicks the trigger
	 *
	 * Receives the image as a dataURL string
	 */
	this.onImage = function () {};

	this.getHueOffset = function ()
	{
		return hueOffset;
	};

	this.setHueOffset = function (h)
	{
		if (fCheckValue(h, "hue"))
			hueOffset = h;
	};

	this.getBrightnessOffset = function ()
	{
		return brightnessOffset;
	};

	this.setBrightnessOffset = function (b)
	{
		if (fCheckValue(b, "brightness"))
			brightnessOffset = b;
	};

	this.getSaturationOffset = function ()
	{
		return saturationOffset;
	};

	this.setSaturationOffset = function (s)
	{
		if (fCheckValue(s, "saturation"))
			saturationOffset = s;
	};
	/**
	 * Closes the videostream, cancels the canvas drawing loop
	 * and frees up the webcam. Use resume()
	 * to continue
	 */
	this.pause = function ()
	{
		if (bIsStopped === false)
		{
			bIsStopped = true;

			if (oStream && oStream.stop)
			{
				oStream.stop();
			}
		}
	};

	/**
	 * Resumes video playback that had previously
	 * been paused with pause().
	 */
	this.resume = function ()
	{
		if (bIsStopped === true)
		{
			bIsStopped = false;
			fRequestWebcamAccess();
		}
	};

	/**
	 * Destroys the photobooth. Closes the video stream,
	 * cancels all outstanding frames and destroys
	 * the DOM elements and eventlisteners.
	 */
	this.destroy = function ()
	{
		this.pause();
		container.removeChild(ePhotobooth);
	};

	/**
	 * Drawing every frame of a video onto
	 * a canvas and performing some pixel
	 * manipulation ( like Photobooth does )
	 * works well in Chrome and Opera, but
	 * crashes Firefox within seconds. Therefore the
	 * Hue / saturation / Brightness sliders are turned
	 * of in FF and the video tag is displayed directly.
	 *
	 * To reeanble the Sliders set this property to true
	 */
	this.forceHSB = false;

	/**
	 * True if the browser supports Webcam streams,
	 * false if not
	 */
	this.isSupported = !!fGetUserMedia;

	/**
	 * Resizes the Photobooth to the desired size.
	 * Minimum for both width and height is 200px
	 */
	this.resize = function (width, height)
	{

		if (height == 0) { // auto-height conserving aspect ratio when height is zero
			height = eVideo.videoHeight / (eVideo.videoWidth / width);

			if (isNaN(height)) {
				//Fix for firefox
				height = width / (4 / 3);
			}
		}

		if (width < 200 || height < 200)
		{
			throw "Error: Not enough space for Photobooth. Min height / width is 200 px";
		}

		_width = width;
		_height = height;
		oResizeHandle.setMax(_width, _height);
		ePhotobooth.style.width = width + "px";
		ePhotobooth.style.height = height + "px";
		eInput.width = width;
		eInput.height = height;
		eOutput.width = width;
		eOutput.height = height;
		eVideo.width = width;
		eVideo.height = height;
	};

	this.capture = function () {
		/**
		 * Flash
		 */
		eBlind.className = "blind";
		eBlind.style.opacity = 1;
		setTimeout(function () {
			eBlind.className = "blind anim";
			eBlind.style.opacity = 0;
		}, 50);

		var mData = {};
		if (oResizeHandle.isActive()) {
			mData = oResizeHandle.getData();
		} else {
			if (bVideoOnly) {
				mData = {
					x: ((_width - eVideo.videoWidth) / 2),
					y: ((_height - eVideo.videoHeight) / 2),
					width: eVideo.videoWidth,
					height: eVideo.videoHeight
				};
			} else {
				mData = {
					x: 0,
					y: 0,
					width: _width,
					height: _height
				};
			}

		}

		var eTempCanvas = cE("canvas");

		eTempCanvas.width = mData.width;
		eTempCanvas.height = mData.height;

		if (bVideoOnly) {
			eTempCanvas.getContext("2d").drawImage(
					eVideo,
					Math.max(0, mData.x - ((_width - eVideo.videoWidth) / 2)),
					Math.max(mData.y - ((_height - eVideo.videoHeight) / 2)),
					mData.width,
					mData.height,
					0,
					0,
					mData.width,
					mData.height);
		} else {
			var oImageData = oOutput.getImageData(mData.x, mData.y, mData.width, mData.height);
			eTempCanvas.getContext("2d").putImageData(oImageData, 0, 0);
		}

		scope.onImage(eTempCanvas.toDataURL());
	};

	/****************************
	 * Private Methods
	 ****************************/

	//include Tools.js
	//include Drag.js
	//include Slider.js
	//include ResizeHandle.js

	var hueOffset = 0,
			saturationOffset = 0,
			brightnessOffset = 0,
			bVideoOnly = false,
			bIsStopped = false,
			oStream = null,
			scope = this,
			_width = container.offsetWidth,
			_height = container.offsetHeight;

	var fCheckValue = function (val, type)
	{
		if (val < -0.5 || val > 0.5)
		{
			throw "Invalid value: " + type + " must be between 0 and 1";
		} else
		{
			return true;
		}
	};

	var fGetAnimFrame =
			(
					window.requestAnimationFrame ||
					window.webkitRequestAnimationFrame ||
					window.mozRequestAnimationFrame ||
					window.oRequestAnimationFrame ||
					window.msRequestAnimationFrame ||
					function (callback) {
						window.setTimeout(callback, 1000 / 60);
					}
			);

	var c = function (n) {
		return ePhotobooth.getElementsByClassName(n)[ 0 ];
	};
	var cE = function (s) {
		return document.createElement(s);
	};

	var ePhotobooth = cE("div");
	ePhotobooth.className = "photobooth";
	ePhotobooth.innerHTML = '<div class="blind"></div><canvas></canvas><div class="warning notSupported">Sorry, Photobooth.js is not supported by your browser</div><div class="warning noWebcam">Please give Photobooth permission to use your Webcam. <span>Try again</span></div><ul><li title="hue"class="hue"></li><li title="saturation"class="saturation"></li><li title="brightness"class="brightness"></li><li title="crop"class="crop"></li><li title="take picture"class="trigger"></li></ul>';

	var eInput = cE("canvas");
	var oInput = eInput.getContext("2d");

	var eOutput = ePhotobooth.getElementsByTagName("canvas")[ 0 ];
	var oOutput = eOutput.getContext("2d");

	var eVideo = cE("video");
	eVideo.autoplay = true;

	var eNoWebcamWarning = c("noWebcam");
	eNoWebcamWarning.getElementsByTagName("span")[ 0 ].onclick = function () {
		fRequestWebcamAccess();
	};

	new Slider(c("hue"), function (value) {
		hueOffset = value;
	});
	new Slider(c("saturation"), function (value) {
		saturationOffset = value;
	});
	new Slider(c("brightness"), function (value) {
		brightnessOffset = value;
	});

	var oResizeHandle = new ResizeHandle(ePhotobooth, _width, _height);

	var eCrop = c("crop");
	eCrop.onclick = function ()
	{
		oResizeHandle.toggle();

		if (eCrop.className === "crop")
		{
			eCrop.className = "crop selected";
		} else
		{
			eCrop.className = "crop";
		}
	};

	var eBlind = c("blind");

	c("trigger").onclick = function ()
	{
		self.capture();
	};


	var fOnStream = function (stream)
	{
		oStream = stream;

		try {
			/**
			 * Chrome
			 */
			eVideo.src = (window.URL || window.webkitURL).createObjectURL(oStream);
			fGetAnimFrame(fNextFrame);
		} catch (e)
		{
			/**
			 * Firefox
			 */
			eVideo.mozSrcObject = oStream;

			if (scope.forceHSB === false)
			{
				bVideoOnly = true;
				ePhotobooth.appendChild(eVideo);
				ePhotobooth.getElementsByTagName("ul")[ 0 ].className = "noHSB";
			} else
			{
				eVideo.addEventListener("canplay", function () {
					fGetAnimFrame(fNextFrame);
				}, false);
			}

			eVideo.play();
		}
	};

	var fOnStreamError = function (e)
	{
		eNoWebcamWarning.style.display = "block";
	};

	var fRequestWebcamAccess = function ()
	{
		eNoWebcamWarning.style.display = "none";
		fGetUserMedia.call(navigator, {"video": true}, fOnStream, fOnStreamError);
	};

	var fHue2rgb = function (p, q, t)
	{
		if (t < 0)
			t += 1;
		if (t > 1)
			t -= 1;
		if (t < 1 / 6)
			return p + (q - p) * 6 * t;
		if (t < 1 / 2)
			return q;
		if (t < 2 / 3)
			return p + (q - p) * (2 / 3 - t) * 6;
		return p;
	};

	var fWrap = function (v)
	{
		if (v > 1)
			return v - 1;
		if (v < 0)
			return 1 + v;
		return v;
	};

	var fCap = function (v)
	{
		if (v > 1)
			return 1;
		if (v < 0)
			return 0;
		return v;
	};

	var fNextFrame = function ()
	{
		try
		{
			oInput.drawImage(eVideo, 0, 0, _width, _height);
		} catch (e) {
		}
		var oImgData = oInput.getImageData(0, 0, _width, _height);
		var pData = oImgData.data;

		for (var i = 0; i < pData.length; i += 4)
		{
			/**
			 * Time for some serious Math. Thanks to MJIJACKSON
			 * for providing a great starting point
			 *
			 * http://mjijackson.com/2008/02/rgb-to-hsl-and-rgb-to-hsv-color-model-conversion-algorithms-in-javascript
			 */
			var r = pData[ i ] / 255,
					g = pData[ i + 1 ] / 255,
					b = pData[ i + 2 ] / 255;

			var max = Math.max(r, g, b),
					min = Math.min(r, g, b);

			var h, s, l = (max + min) / 2;

			if (max == min)
			{
				h = s = 0; // achromatic
			} else
			{
				var d = max - min;
				s = l > 0.5 ? d / (2 - max - min) : d / (max + min);

				if (max === r)
					h = ((g - b) / d + (g < b ? 6 : 0)) / 6;
				if (max === g)
					h = ((b - r) / d + 2) / 6;
				if (max === b)
					h = ((r - g) / d + 4) / 6;
			}

			h = fWrap(h + hueOffset);
			s = fCap(s + saturationOffset);
			l = fCap(l + brightnessOffset);

			if (s === 0)
			{
				r = g = b = l; // achromatic
			} else
			{
				var q = l < 0.5 ? l * (1 + s) : l + s - l * s;
				var p = 2 * l - q;
				r = fHue2rgb(p, q, h + 1 / 3);
				g = fHue2rgb(p, q, h);
				b = fHue2rgb(p, q, h - 1 / 3);
			}

			pData[ i ] = r * 255;
			pData[ i + 1 ] = g * 255;
			pData[ i + 2 ] = b * 255;
		}

		oOutput.putImageData(oImgData, 0, 0);

		if (bIsStopped === false)
		{
			fGetAnimFrame(fNextFrame);
		}

	};

	this.resize(_width, _height);
	container.appendChild(ePhotobooth);

	if (fGetUserMedia)
	{
		/**
		 * Startup
		 */
		fGetAnimFrame(fRequestWebcamAccess);
	} else
	{
		c("notSupported").style.display = "block";
	}
};